### ISSUE TITLE
Wrong output dimension calculation of StrideSlice operation on TensorFlow Lite.

### ISSUE THREAD

**Issue by kukdh1** [tensorflow/tensorflow#46092]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS 8
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Both pip and source build
- TensorFlow version (use command below): `v2.4.0-rc4-71-g582c8d236cb 2.4.0`
- Python version: `3.6.8`
- Bazel version (if compiling from source): `3.7.1`
- GCC/Compiler version (if compiling from source): `8.3.1`
- CUDA/cuDNN version: CUDA 11.1 / cuDNN 8
- GPU model and memory: RTX3090 24GB

**Describe the current behavior**
The `StridedSlice` operation on TensorFlow Lite calculates output dimension incorrectly.

I'm developing some custom operations for both TensorFlow and TensorFlow Lite.
While I debugging my custom operation, inference fails only on TensorFlow Lite.
I found that `StridedSlice` operation of TensorFlow Lite calculates output dimension incorrectly.
I added some `printf` function to [`tflite::ops::builtin::strided_slice::Eval`](https://github.com/tensorflow/tensorflow/blob/582c8d236cb079023657287c318ff26adb239002/tensorflow/lite/kernels/strided_slice.cc#L185):
```c++
TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
  StridedSliceContext op_context(context, node);

  printf("**** [STRIDED_SLICE] Input [");
  for (int i = 0; i < NumDimensions(op_context.input); i++) {
    printf("%d, ", SizeOfDimension(op_context.input, i));
  }
  printf("]\n**** [STRIDED_SLICE] Begin: ");
  for (int i = 0; i < SizeOfDimension(op_context.begin, 0); i++) {
    printf("%d, ", GetTensorData<int32_t>(op_context.begin)[i]);
  }
  printf("\n**** [STRIDED_SLICE] End: ");
  for (int i = 0; i < SizeOfDimension(op_context.end, 0); i++) {
    printf("%d, ", GetTensorData<int32_t>(op_context.end)[i]);
  }
  printf("\n**** [STRIDED_SLICE] Strides: ");
  for (int i = 0; i < SizeOfDimension(op_context.strides, 0); i++) {
    printf("%d, ", GetTensorData<int32_t>(op_context.strides)[i]);
  }
  printf("\n");

  if (IsDynamicTensor(op_context.output)) {
    TF_LITE_ENSURE_OK(context, ResizeOutputTensor(context, &op_context));
  }
  StridedSliceParams op_params = BuildStridedSliceParams(&op_context);

  printf("**** [STRIDED_SLICE] Output [");
  for (int i = 0; i < NumDimensions(op_context.output); i++) {
    printf("%d, ", SizeOfDimension(op_context.output, i));
  }
  printf("]\n");
```

And I got following logs:
```
**** [STRIDED_SLICE] Input [1656, 8, 32, ]
**** [STRIDED_SLICE] Begin: 0, 0, 0, 
**** [STRIDED_SLICE] End: 893, 
**** [STRIDED_SLICE] Strides: 1, 1, 1, 
**** [STRIDED_SLICE] Output [893, 8, 0, ]
```

**Describe the expected behavior**
After strided slicing, output dimension should be `[893, 8, 32, ]` in above example.

**Standalone code to reproduce the issue**
I wrote a smallest [reproducible code](https://github.com/kukdh1/tflite_buf_report/blob/master/test_strided_slice.py).
The `SimpleLayer` keras layer creates read-only `Tensor` with size of [128, 8, 32].
After ten `model.call`s, I saved the model into tflite model file.

On TensorFlow (python, no code modification -- installed by pip `tensorflow-2.4.0-cp36-cp36m-manylinux2010_x86_64.whl`) it calculates output dimension correctly.
```
TEST: Input: [128, 8, 32] SiliceTo: 53 Output: (53, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 38 Output: (38, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 64 Output: (64, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 100 Output: (100, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 106 Output: (106, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 90 Output: (90, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 126 Output: (126, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 122 Output: (122, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 62 Output: (62, 8, 32)
TEST: Input: [128, 8, 32] SiliceTo: 99 Output: (99, 8, 32)
```

I checked the stored model with [netron](https://netron.app/):
![image](https://user-images.githubusercontent.com/6904750/103407030-52be8400-4ba0-11eb-8344-2e2d1b85e0d2.png)
And it shows that `StridedSlice` has input `Tensor` of size `[128, 8, 32]` and output `Tensor` has size `[Unknown, 8, 32]`.

To test on TensorFlow Lite, I wrote [simple inference code](https://github.com/kukdh1/tflite_buf_report/blob/master/strided_slice.cc) with TensorFlow Lite C++ API.
It randomly selects input `Tensor` (slice index) and print the output dimension.

I used the source code (582c8d23 == v2.4.0) with above `StrideSlice` `printf` modification.
I built `libtensorflowlite.so` with following command:
`bazel build --verbose_failures -c opt --define=no_aws_support=true --define=no_gcp_support=true --define=no_hdfs_support=true --define=no_nccl_support=true --define=build_with_mkl=false --config=monolithic //tensorflow/lite:libtensorflowlite.so`
And the content of `.tf_configure.bazelrc` is:
```
build --action_env PYTHON_BIN_PATH="/home/kukdh1/.virtualenvs/tf_develop/bin/python3"
build --action_env PYTHON_LIB_PATH="/home/kukdh1/.virtualenvs/tf_develop/lib/python3.6/site-packages"
build --python_path="/home/kukdh1/.virtualenvs/tf_develop/bin/python3"
build --config=xla
build --action_env CUDA_TOOLKIT_PATH="/usr/local/cuda-11.1"
build --action_env TF_CUDA_COMPUTE_CAPABILITIES="7.5,8.6"
build --action_env LD_LIBRARY_PATH="/usr/local/cuda-11.1/lib64:"
build --action_env GCC_HOST_COMPILER_PATH="/usr/bin/gcc"
build --config=cuda
build:opt --copt=-march=native
build:opt --copt=-Wno-sign-compare
build:opt --host_copt=-march=native
build:opt --define with_default_optimizations=true
test --flaky_test_attempts=3
test --test_size_filters=small,medium
test --test_env=LD_LIBRARY_PATH
test:v1 --test_tag_filters=-benchmark-test,-no_oss,-no_gpu,-oss_serial
test:v1 --build_tag_filters=-benchmark-test,-no_oss,-no_gpu
test:v2 --test_tag_filters=-benchmark-test,-no_oss,-no_gpu,-oss_serial,-v1only
test:v2 --build_tag_filters=-benchmark-test,-no_oss,-no_gpu,-v1only
build --action_env TF_CONFIGURE_IOS="0"
```

I built test program with `mkdir build && cmake -DTF_SOURCE_DIR=<source directory> ..` ([CMakeLists.txt](https://github.com/kukdh1/tflite_buf_report/blob/master/CMakeLists.txt))
When I ran the test program, I got following result:
```
$ CUDA_VISIBLE_DEVICES=-1 ./tflite-strided-slice strided_slice.tflite 
2020-12-31 19:47:18.088431: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
**** [STRIDED_SLICE] Input [128, 8, 32, ]
**** [STRIDED_SLICE] Begin: 0, 0, 0, 
**** [STRIDED_SLICE] End: 103, 
**** [STRIDED_SLICE] Strides: 1, 1, 1, 
**** [STRIDED_SLICE] Output [103, 0, 32, ]
Input: 103
Output: [103 0 32 ]
########## PRINT INTERPRETER STATE BEGIN ##########
Interpreter has 6 tensors and 2 nodes
Inputs: 0
Outputs: 5

Tensor   0 input_1              kTfLiteInt32   kTfLiteCustom          4 bytes ( 0.0 MB) 
Tensor   1 simple_layer/ReadVariableOp/resource kTfLiteFloat32   kTfLiteMmapRo     131072 bytes ( 0.1 MB)  128 8 32
Tensor   2 simple_layer/strided_slice kTfLiteInt32   kTfLiteMmapRo         12 bytes ( 0.0 MB)  3
Tensor   3 simple_layer/strided_slice1 kTfLiteInt32   kTfLiteMmapRo         12 bytes ( 0.0 MB)  3
Tensor   4 simple_layer/strided_slice/stack_1 kTfLiteInt32  kTfLiteArenaRw          4 bytes ( 0.0 MB)  1
Tensor   5 Identity             kTfLiteFloat32  kTfLiteDynamic          0 bytes ( 0.0 MB)  103 0 32

Node   0 Operator Builtin Code  83 PACK
  Inputs: 0
  Outputs: 4
Node   1 Operator Builtin Code  45 STRIDED_SLICE
  Inputs: 1 2 4 3
  Outputs: 5
########### PRINT INTERPRETER STATE END ###########
```

The output Tensor has size of `[103, 0, 32]` not `[103, 8, 32]`.

I think this is a bug on [`tflite::ops::builtin::strided_slice::ResizeOutputTensor`](https://github.com/tensorflow/tensorflow/blob/582c8d236cb079023657287c318ff26adb239002/tensorflow/lite/kernels/strided_slice.cc#L101).
Please let me know if I did something wrong.

**Other info / logs**
You can find all the code snippets I used to reproduce the problem at [here](https://github.com/kukdh1/tflite_buf_report).
(Download all dependencies by running `./download_dependencies.sh` at `tensorflow/lite/tools/make`)
You can find tflite model too.

P.S. The dimension error is quite random. Sometimes it calculates correctly. Sometimes innermost (axis=2) dimension becomes zero. Sometimes middle (axis=1) dimension becoms zero, Sometimes both (axis=1, axis=2) dimensions become zero.

P.S. 2. I double checked with unmodified tensorflow source downloaded with:
`wget https://github.com/tensorflow/tensorflow/archive/v2.4.0.tar.gz`
on Ubuntu 18.04.5 (different machine) with GCC 7.5.0, bazel 3.7.2 and Python 3.6.9.
Same command used to build `libtensorflowlite.so` and `.tf_configure.bazelrc` is:
```
build --action_env PYTHON_BIN_PATH="/home/kukdh1/.virtualenvs/tensorflow/bin/python3"
build --action_env PYTHON_LIB_PATH="/home/kukdh1/.virtualenvs/tensorflow/lib/python3.6/site-packages"
build --python_path="/home/kukdh1/.virtualenvs/tensorflow/bin/python3"
build --config=xla
build:opt --copt=-march=native
build:opt --copt=-Wno-sign-compare
build:opt --host_copt=-march=native
build:opt --define with_default_optimizations=true
test --flaky_test_attempts=3
test --test_size_filters=small,medium
test:v1 --test_tag_filters=-benchmark-test,-no_oss,-gpu,-oss_serial
test:v1 --build_tag_filters=-benchmark-test,-no_oss,-gpu
test:v2 --test_tag_filters=-benchmark-test,-no_oss,-gpu,-oss_serial,-v1only
test:v2 --build_tag_filters=-benchmark-test,-no_oss,-gpu,-v1only
build --action_env TF_CONFIGURE_IOS="0"
```

The output on Ubuntu machine was (one example):
```
Input: 41
Output: [41 0 0 ]
########## PRINT INTERPRETER STATE BEGIN ##########
Interpreter has 6 tensors and 2 nodes
Inputs: 0
Outputs: 5

Tensor   0 input_1              kTfLiteInt32   kTfLiteCustom          4 bytes ( 0.0 MB)
Tensor   1 simple_layer/ReadVariableOp/resource kTfLiteFloat32   kTfLiteMmapRo     131072 bytes ( 0.1 MB)  128 8 32
Tensor   2 simple_layer/strided_slice kTfLiteInt32   kTfLiteMmapRo         12 bytes ( 0.0 MB)  3
Tensor   3 simple_layer/strided_slice1 kTfLiteInt32   kTfLiteMmapRo         12 bytes ( 0.0 MB)  3
Tensor   4 simple_layer/strided_slice/stack_1 kTfLiteInt32  kTfLiteArenaRw          4 bytes ( 0.0 MB)  1
Tensor   5 Identity             kTfLiteFloat32  kTfLiteDynamic          0 bytes ( 0.0 MB)  41 0 0

Node   0 Operator Builtin Code  83 PACK
  Inputs: 0
  Outputs: 4
Node   1 Operator Builtin Code  45 STRIDED_SLICE
  Inputs: 1 2 4 3
  Outputs: 5
########### PRINT INTERPRETER STATE END ###########
```

**Comment by kukdh1** [comment]
Hi,

I found the function `BuildStridedSliceParams` in `tensorflow/lite/kernels/strided_slice.cc` was the problem.
The following loop assumes the dimension of `begin`, `end` and `strides` are same, but in my case, dimension of `begin` and `strides` was `3` while dimension of `end` was `1`.
```cpp
  int begin_count = GetTensorShape(op_context->begin).Dims(0);
  for (int i = 0; i < begin_count; ++i) {
    op_params.start_indices[i] = GetTensorData<int32_t>(op_context->begin)[i];
    op_params.stop_indices[i] = GetTensorData<int32_t>(op_context->end)[i];
    op_params.strides[i] = GetTensorData<int32_t>(op_context->strides)[i];
  }
```

So I changed the code to:
```diff
diff --git a/tensorflow/lite/kernels/strided_slice.cc b/tensorflow/lite/kernels/strided_slice.cc
index d10e99c1..26f3ab5c 100644
--- a/tensorflow/lite/kernels/strided_slice.cc
+++ b/tensorflow/lite/kernels/strided_slice.cc
@@ -78,19 +78,29 @@ StridedSliceParams BuildStridedSliceParams(StridedSliceContext* op_context) {
   op_params.shrink_axis_mask = op_context->params->shrink_axis_mask;
 
   int begin_count = GetTensorShape(op_context->begin).Dims(0);
-  for (int i = 0; i < begin_count; ++i) {
-    op_params.start_indices[i] = GetTensorData<int32_t>(op_context->begin)[i];
-    op_params.stop_indices[i] = GetTensorData<int32_t>(op_context->end)[i];
-    op_params.strides[i] = GetTensorData<int32_t>(op_context->strides)[i];
-  }
+  int end_count = GetTensorShape(op_context->end).Dims(0);
+  int strides_count = GetTensorShape(op_context->strides).Dims(0);
+
+  for (int i = 0; i < op_context->dims; ++i) {
+    if (i < begin_count) {
+      op_params.start_indices[i] = GetTensorData<int32_t>(op_context->begin)[i];
+    } else {
+      op_params.start_indices[i] = 0;
+      op_params.begin_mask |= (1 << i);
+    }
 
-  // If the length of begin and end smaller than number of input dims, set the
-  // mask bit of begin and end for that index.
-  for (int i = begin_count; i < op_context->dims; ++i) {
-    op_params.start_indices[i] = op_params.stop_indices[i] = 0;
-    op_params.strides[i] = 1;
-    op_params.begin_mask |= (1 << i);
-    op_params.end_mask |= (1 << i);
+    if (i < end_count) {
+      op_params.stop_indices[i] = GetTensorData<int32_t>(op_context->end)[i];
+    } else {
+      op_params.stop_indices[i] = 0;
+      op_params.end_mask |= (1 << i);
+    }
+
+    if (i < strides_count) {
+      op_params.strides[i] = GetTensorData<int32_t>(op_context->strides)[i];
+    } else {
+      op_params.strides[i] = 1;
+    }
   }
   return op_params;
 }
```

And it solves the problem.
(In below interpreter state, `Tensor 2 == begin` and `Tensor 3 == strides` has dimension of `3`, but `Tensor 4 == end` has dimension of `1`).

```
Input: 41
Output: [41 8 32 ]
########## PRINT INTERPRETER STATE BEGIN ##########
Interpreter has 6 tensors and 2 nodes
Inputs: 0
Outputs: 5

Tensor   0 input_1              kTfLiteInt32   kTfLiteCustom          4 bytes ( 0.0 MB) 
Tensor   1 simple_layer/ReadVariableOp/resource kTfLiteFloat32   kTfLiteMmapRo     131072 bytes ( 0.1 MB)  128 8 32
Tensor   2 simple_layer/strided_slice kTfLiteInt32   kTfLiteMmapRo         12 bytes ( 0.0 MB)  3
Tensor   3 simple_layer/strided_slice1 kTfLiteInt32   kTfLiteMmapRo         12 bytes ( 0.0 MB)  3
Tensor   4 simple_layer/strided_slice/stack_1 kTfLiteInt32  kTfLiteArenaRw          4 bytes ( 0.0 MB)  1
Tensor   5 Identity             kTfLiteFloat32  kTfLiteDynamic      41984 bytes ( 0.0 MB)  41 8 32

Node   0 Operator Builtin Code  83 PACK
  Inputs: 0
  Outputs: 4
Node   1 Operator Builtin Code  45 STRIDED_SLICE
  Inputs: 1 2 4 3
  Outputs: 5
########### PRINT INTERPRETER STATE END ###########
```

This patch solves the problem, but I'm not sure that fixing `BuildStridedSliceParams` function is correct solution.
Maybe, in conversion process, all three (`begin`, `end` and `strides`) tensors should not expanded or should be expanded.
(I used MLIR for the conversion)

Thanks.

**Comment by WindQAQ** [comment]
Hi @kukdh1, conversion as well as output seems correct in tf-nightly. Can you try it on nightly version instead?

![image](https://user-images.githubusercontent.com/11615393/103586386-d5458d00-4e99-11eb-922a-2261aa38005b.png)

**Comment by kukdh1** [comment]
Hi,

I just installed `tf_nightly-2.5.0.dev20210104-cp36-cp36m-manylinux2010_x86_64.whl` and the converted result has correct input dimensions.
![image](https://user-images.githubusercontent.com/6904750/103616503-e95bbe00-4f6f-11eb-96f9-1a8601c7caad.png)

And I compiled `libtensorflowlite.so` with master branch bd68b49a and it calculates output dimension as well.
```
Input: 103
Output: [103 8 32 ]
########## PRINT INTERPRETER STATE BEGIN ##########
Interpreter has 6 tensors and 2 nodes
Inputs: 0
Outputs: 5

Tensor   0 input_1              kTfLiteInt32   kTfLiteCustom          4 bytes ( 0.0 MB)
Tensor   1 simple_layer/ReadVariableOp/resource kTfLiteFloat32   kTfLiteMmapRo     131072 bytes ( 0.1 MB)  128 8 32
Tensor   2 simple_layer/strided_slice/stack kTfLiteInt32   kTfLiteMmapRo          4 bytes ( 0.0 MB)  1
Tensor   3 simple_layer/strided_slice/stack_2 kTfLiteInt32   kTfLiteMmapRo          4 bytes ( 0.0 MB)  1
Tensor   4 simple_layer/strided_slice/stack_1 kTfLiteInt32  kTfLiteArenaRw          4 bytes ( 0.0 MB)  1
Tensor   5 Identity             kTfLiteFloat32  kTfLiteDynamic     105472 bytes ( 0.1 MB)  103 8 32

Node   0 Operator Builtin Code  83 PACK
  Inputs: 0
  Outputs: 4
Node   1 Operator Builtin Code  45 STRIDED_SLICE
  Inputs: 1 2 4 3
  Outputs: 5
########### PRINT INTERPRETER STATE END ###########
```

I hope this will be fixed in v2.4.1.

Thanks.

**Comment by Saduf2019** [comment]
@kukdh1 
Please move the issue to closed status if resolved.
